homeassistant/components/powerwall/entity.py

```diff
@@ -43,3 +44,49 @@ def __init__(self, powerwall_data: PowerwallRuntimeData) -> None:
     def data(self) -> PowerwallData:
         """Return the coordinator data."""
         return self.coordinator.data
+
+
+class BatteryEntity(CoordinatorEntity[DataUpdateCoordinator[PowerwallData]]):
+    """Base class for battery entities."""
+
+    def __init__(
+        self, powerwall_data: PowerwallRuntimeData, serial_number: str
+    ) -> None:
+        """Initialize the entity."""
+        base_info = powerwall_data[POWERWALL_BASE_INFO]
+        coordinator = powerwall_data[POWERWALL_COORDINATOR]
+        assert coordinator is not None
+        super().__init__(coordinator)
+        self.serial_number = serial_number
+        self.power_wall = powerwall_data[POWERWALL_API]
+        self.base_unique_id = f"{base_info.gateway_din}_{serial_number}"
+
+        battery: Optional[BatteryResponse] = None
+        for b in base_info.batteries:
+            if b.serial_number == serial_number:
+                battery = b
+        assert battery is not None
```

Comment:
Looking for a better way to handle runtime issues like this...

Reply:
Since you already have the battery object, pass it in here instead of the serial_number